backend/src/main/java/com/yohoard/notes/TagService.java

```diff
@@ -16,24 +15,22 @@ public TagService(TagRepository tagRepository) {
         this.tagRepository = tagRepository;
     }
 
-    public List<Tag> getOrCreate(List<String> names) {
-        List<Tag> existingTags = tagRepository.findByNameIn(names);
+    public List<Tag> getOrCreate(Set<String> names) {
```

Comment:
Chodziło mi raczej o coś takiego:

```
public List<Tag> getOrCreate(Set<String> names) {
        List<Tag> tags = tagRepository.findByNameIn(names);

        Set<Tag> newTags = names.stream()
            .filter(name -> tags.stream().noneMatch(tag -> tag.getName().equals(name)))
            .map(name -> new Tag(name, "", 0))
            .collect(Collectors.toUnmodifiableSet());
        
        tagRepository.saveAll(newTags);

        tags.addAll(newTags);

        return tags;
    }
```

To co podaje user wchodzi w parametrze jako Set<String> names, trzeba to porównać z tym co już jest w bazie, czyli List<Tag> tags, i to czego brakuje stworzyć. Plus dodałam .toUnmodifiableSet(), bo w tym miejscu nie jest potrzebna mutowalna kolekcja, skoro to leci potem do bazy i dokleja się do listy.

Jest tu jeszcze trochę miejsca na refaktor, ale to lepiej robić jak będą testy 😛